### PR TITLE
[reissue] kubeadm: Split out ClusterConfiguration from InitConfiguration

### PR DESCRIPTION
**Pull request by rosti**
As @luxas is not able to take care of #66219, I am reissuing the same change here. There are a few minor things added by me:

- The original PR is rebased on latest master.
- Some broken tests were fixed.
- Some TODOs were added.
- Run update-bazel and update-gofmt

Below is the text of the original PR by Lucas.

-----

**What this PR does / why we need it:**

Splits MasterConfiguration to InitConfiguration and ClusterConfiguration as outlined in the kubeadm Config KEP. InitConfiguration holds init-only information, and ClusterConfiguration holds cluster-wide information. In the internal representation InitConfiguration wraps ClusterConfiguration as a field, but in serialized format they're different YAML documents.

**Which issue(s) this PR fixes** (optional, in fixes #<issue number>(, fixes #<issue_number>, ...) format, will close the issue(s) when PR gets merged):
ref: kubernetes/kubeadm#911
Depends on:

- [X] #65776
- [X] #65628
- [X] #65629
- [X] #65631
- [X] #65940
- [X] #65787
- [X] #65945
- [X] #65951
- [X] #65952

**Special notes for your reviewer:**

**Release note**:
```release-note
kubeadm: InitConfiguration now consists of two structs: InitConfiguration and ClusterConfiguration
```

@kubernetes/sig-cluster-lifecycle-pr-reviews

### CODE REVIEW

**Comment by neolit123** [comment]
thanks for doing this PR!

it needs a:
```
./hack/update-gofmt.sh
./hack/update-bazel.sh
```

but other than that i think we should merge ASAP and apply changes on top if needed.

**Review by fabriziopandini — COMMENTED** [review]
@rosti +100 for jumping in and helping on @luxas work!

/approve

Being this so big, I kindly ask @timothysc or @liztio (or *) to give a second pass + final lgtm

**Comment by chuckha** [comment]
this code looks good to me. I'd rather get this merged ASAP as @neolit123 points out and fix anything that breaks before the release

File: cmd/kubeadm/app/apis/kubeadm/fuzzer/fuzzer_test.go [hunk 1]
```diff
@@ -16,7 +16,12 @@ limitations under the License.
 
 package fuzzer
 
-import (
+// TODO: Fuzzing rouudtrip tests are currently disabled in the v1.12 cycle due to the
+// v1alpha2 -> v1alpha3 migration. As the ComponentConfigs were embedded in the structs
```
Comment (timothysc) [inline cmd/kubeadm/app/apis/kubeadm/fuzzer/fuzzer_test.go]: Can you open an issue here?

File: cmd/kubeadm/app/cmd/config.go [hunk 2]
```diff
@@ -160,19 +163,31 @@ func getAllAPIObjectNames() []string {
 
 func getDefaultedInitConfig() (*kubeadmapi.InitConfiguration, error) {
 	return configutil.ConfigFileAndDefaultsToInternalConfig("", &kubeadmapiv1alpha3.InitConfiguration{
-		API:               kubeadmapiv1alpha3.API{AdvertiseAddress: "1.2.3.4"},
-		BootstrapTokens:   []kubeadmapiv1alpha3.BootstrapToken{sillyToken},
-		KubernetesVersion: fmt.Sprintf("v1.%d.0", constants.MinimumControlPlaneVersion.Minor()+1),
+		// TODO: Probably move to getDefaultedClusterConfig?
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/config.go]: ?

**Review by timothysc — COMMENTED** [review]
What's with all the TODOs?  are we going to follow on or address here?  

@fabriziopandini - I've been out for too long so I've lost state.

**Comment by timothysc** [comment]
@rosti - needs a rebase 

/lgtm

**Comment by rosti** [comment]
Rebased the PR.

File: cmd/kubeadm/app/cmd/config.go (hunk 2 above)
Comment (rosti) [reply cmd/kubeadm/app/cmd/config.go]: I think, this one was a comment in the old PR. We decided to mark all unaddressed comments in the original PR as TODOs in the new one.

**Pull request merged**